### PR TITLE
Add openstack/auth_yaml.go

### PR DESCRIPTION
**Pull request by raphink**
This aims to fix https://github.com/rackspace/gophercloud/issues/487.

/cc @emonty @mcanevet

for #104

### CODE REVIEW

**Comment by raphink** [comment]
@jtopjian @jrperritt is that good for you?

**Comment by jtopjian** [comment]
Hi @raphink! I think just as a formality, there should be a corresponding issue opened for this. Check the [style guide](https://github.com/gophercloud/gophercloud/blob/master/STYLEGUIDE.md) for more information.

As for code review, I'll leave that one for @jrperritt. :)

**Comment by raphink** [comment]
Hi @jtopjian . There was an issue on the rackspace project: https://github.com/rackspace/gophercloud/issues/487

Should this issue be copied to the gophercloud project?

**Comment by raphink** [comment]
There https://github.com/gophercloud/gophercloud/issues/104

File: openstack/auth_yaml.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+		Username:         username,
+		Password:         password,
+		TenantID:         tenantID,
+		TenantName:       tenantName,
+	}
+
+	return ao, nil
+}
+
+func cloudsYaml() ([]byte, error) {
```
Comment (jrperritt) [inline openstack/auth_yaml.go]: I understand that `os-client-config` has logic that looks in all these different places and decides what to do for the user, but I think we should just have the user tell us the file location:
```go
func AuthOptionsFromYaml(yamlPath, cloud string) (gophercloud.AuthOptions, error) {
...
}
```

File: openstack/auth_yaml.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+	auth := clouds.Clouds[cloud].Auth
+
+	authURL := auth.AuthURL
+	username := auth.Username
+	password := auth.Password
+	tenantID := auth.ProjectID
+	tenantName := auth.ProjectName
+
+	if env {
```
Comment (raphink) [inline openstack/auth_yaml.go]: This provides overrides through the environment, which is the default behavior usually. If this isn't done here, it makes it much harder to implement, as you'll have to call both `AuthOptionsFromYaml` and `AuthOptionsFromEnv` and merge the keys one by one.

File: openstack/auth_yaml.go (hunk 1 above)
Comment (raphink) [inline openstack/auth_yaml.go]: This means every client will need to reimplement the default locations for this file, even though they are listed in the spec.

Comment (jrperritt) [inline openstack/auth_yaml.go]: Presumably a user has only one config file. It seems like letting them explicitly specify the location and name of the file is what we'd want.

File: openstack/auth_yaml.go (hunk 2 above)
Comment (jrperritt) [inline openstack/auth_yaml.go]: I don't see this as a common situation. Moreover, a user could just:
```go
ao, err := openstack.AuthOptionsFromYaml("config.yaml", "myCloud")
ao.Username = os.Getenv("OS_USERNAME")
```

**Comment by raphink** [comment]
OK, I guess I'll move the getenv + file loop logic to terraform then...

File: openstack/auth_yaml.go (hunk 2 above)
Comment (raphink) [inline openstack/auth_yaml.go]: Another reason why I'd done this is the checks, e.g.:

```go
	if authURL == "" {
		err := gophercloud.ErrMissingInput{Argument: "authURL"}
		return nilOptions, err
	}
```

It's OK to not have variables set in the YAML files, so long as they are overridden in the environment. If I remove the environment option, then I'll have to remove the checks, too.

File: openstack/auth_yaml.go (hunk 1 above)
Comment (raphink) [inline openstack/auth_yaml.go]: So you'd want another parameter to the method for specifying the file to parse.

**Comment by mcanevet** [comment]
@jtopjian @jrperritt is there something preventing this PR to be merged?

**Comment by jtopjian** [comment]
I haven't been following this one, but it looks like the hold is due to more discussion being needed on some parts.

From reading the comments, I can see both sides. I'm hesitant to give an opinion because I have no clue how to design software correctly, but:

Gophercloud isn't really a full-fledged client. It's a core library. The packages for communicating with the various APIs are thin. There's some convenience here and there (detecting environment variables for auth, the various `IDFromName` functions), but it's mostly the bare minimum needed to communicate with an API. I've grown to appreciate this very much.

IMO, this feature feels slightly more convenient than core, so I can understand the pushback in some areas. For example, instead of checking a bunch of different directories for the config file, it's much simpler to specify the file path. Then Gophercloud never has to worry about updating default paths. 

Yes, that means that everything that consumes Gophercloud will have to re-implement those same types of checks. For now. I suspect this is why https://github.com/openstack/os-client-config is actually a separate project rather than part of either OSC or Keystone, right?

Maybe this function is better suited for something discussed [here](https://github.com/gophercloud/gophercloud/issues/237) than in Gophercloud directly? In the end, you'd still end up passing values to the core AuthOptions. And tools like Terraform could implement the "utility" package rather than rewriting the implementation.

I could be totally wrong, though.

**Comment by jtopjian** [comment]
Hi all,

A new [utils](https://github.com/gophercloud/utils) repo was created a few weeks ago. I think this feature might be better suited there. Creating a draft of this in the utils repo is on my to-do list, which I will _eventually_ get to, but I wanted to leave a note here in case someone else wanted to submit it first. I would be happy to review.

**Comment by jtopjian** [comment]
Hi all,

https://github.com/gophercloud/utils/pull/2 has just been merged which adds initial support for `os-client-config`-like functionality, including `clouds.yaml`. I don't have an in-depth amount of knowledge of `os-client-config`, so I anticipate it requiring some follow-up patches. Please don't hesitate to patch or submit an issue.

I'm going to go ahead and close this.

**Pull request closed without merging**